### PR TITLE
Speed up .edf export with `edfio`

### LINKED ISSUE #9883: Support STIM channels in EDF export
I found some issues with our recently added EDF export that we might want to fix before the next release:

1. The error if `EDFlib-Python` is not installed is misleading, because it says `RuntimeError: For exporting to EDF to work, the EDFlib module is needed, but it could not be imported`. However, although the import name is `EDFlib`, the PyPI name is actually `EDFlib-Python` and `EDFlib` is a different PyPI package (which requires Cython etc.).
2. `EDFlib-Python` is super slow compared to `pyEDFlib` (which uses the C implementation). I tested exporting with one of our BDF files from our lab (425MB), and it took `EDFlib-Python` 1 minute and 52 seconds to export. For comparison, `pyEDFlib` took only 6 seconds.
3. Analog stim channels are not exported. This is essential in BDF files, because Biosemi amplifiers do not have discrete triggers and instead always write an analog stim channel. There is no technical reason to not export stim channels in addition to EEG (other than carefully setting the scaling, unit, and physical min/max). I've fixed this in my EDF/BDF export implementation in MNELAB recently (today) in https://github.com/cbrnr/mnelab/pull/230.

The first issue is probably easy to fix. The second issue is pretty serious, we can't expect that users wait almost 2 minutes for something that can be done in 6 seconds. I'd consider switching to `pyEDFlib` instead (they have binary wheels and AFAIR the restriction to use pure Python applies only to our project but not external dependencies). The third issue requires some work, but this is important for BDF files.

### PR DESCRIPTION
**Pull request by hofaflo**
#### Reference issue
Partially addresses #9883 (first and second issue), might supersede #12085

#### What does this implement/fix?
As suggested by @cbrnr [here](https://github.com/mne-tools/mne-python/pull/12085#issuecomment-1803605755), [edfio](https://github.com/the-siesta-group/edfio) could be used to speed up the export to .edf files.

#### Additional information
- This also fixes a bug where sometimes unrelated channels were used to determine the physical min and max for `physical_range="auto"` (see [here](https://github.com/hofaflo/mne-python/blob/b3f4efc0129f87892b006cd71560696256572042/mne/export/_edf.py#L110)).
- Two sections of `test_rawarray_edf` are removed, because
  - the patient birth date is formatted as "dd-MMM-yyyy" in EDF+, so a date like 20-JAN-1700 is actually valid
  - the export always uses EDF+C, which allows to store dates outside the range valid in "original" Edf (1985-01-01 to 2084-12-31) formatted as "dd-MMM-yyyy" in the "local recording identification" field

### CODE REVIEW

**Comment by hofaflo** [comment]
Conda+mamba checks fail since edfio is not on conda-forge yet, I'll work on that.

**Comment by larsoner** [comment]
> edfio is not on conda-forge yet, I'll work on that.

Let me know if you need help, I did this for `pysnirf2` just yesterday so it's fresh in my mind. But the short version is something like:

1. clone `staged-recipes`
2. `pip install grayskull` (or conda install it)
3. cd to `staged-recipes/recipes`
4. `grayskull pypi edfio`
5. Open PR to `staged-recipes`

**Comment by hofaflo** [comment]
Thank you for the simplified instructions!
Review is pending at conda-forge/staged-recipes#24511

**Comment by Teuniz** [comment]
>     * Two sections of `test_rawarray_edf` are removed, because
>       
>       * the patient birth date is formatted as "dd-MMM-yyyy" in EDF+, so a date like 20-JAN-1700 is actually valid
>       * the export always uses EDF+C, which allows to store dates outside the range valid in "original" Edf (1985-01-01 to 2084-12-31) formatted as "dd-MMM-yyyy" in the "local recording identification" field

Pay attention. Allowing a measurement date outside the range 1985-01-01 to 2084-12-31 breaks backwards compatibility with EDF!
In order to protect this backwards compatibility, EDFbrowser will refuse to open an EDF+ file that has a startdate like, for example, 1700-01-01 or 2090-01-01.
(Ofcourse, this limitation in EDFbrowser will be removed before the year 2085 arrives.)

**Comment by hofaflo** [comment]
Thank you for the heads up @Teuniz! edfio currently stores non-EDF-compatible startdates according to EDF+ specs in the "local recording identification" field and sets the EDF startdate field to "01.01.85". On read, it prefers the EDF+ startdate to the EDF startdate, but of course that only makes sense if no other software is used. Since EDFbrowser does not open files with such a mismatch, we will update edfio to refuse writing them :)

**Comment by Teuniz** [comment]
> Since EDFbrowser does not open files with such a mismatch, we will update edfio to refuse writing them :)

Thanks! In an "ideal" world where all EDF software has been updated to EDF+, this shouldn't be necessary,
but ofcourse this is not the case.
The goal is to avoid confusion when two different viewers (one EDF, the other EDF+) show different startdates.
Thanks again for your consideration. It's much appreciated.

**Comment by hofaflo** [comment]
[Conda installs](https://anaconda.org/conda-forge/edfio) should work now, but I'll mark this as draft until the-siesta-group/edfio#5 is merged and released.

**Comment by sappelhoff** [comment]
note: once this is merged, we'll have to make adjustments in mne-bids:

https://github.com/sappelhoff/mne-bids/blob/781fc577dc0da0419f7eb5c40d537b51834cb802/doc/install.rst?plain=1#L27

**Comment by cbrnr** [comment]
Just to be clear, even EDF+ mandates that the date range is between 1985–2084 (https://www.edfplus.info/specs/edfplus.html#additionalspecs). So regardless of EDFbrowser not opening it, this is non-standard and should therefore not be supported.

> (Ofcourse, this limitation in EDFbrowser will be removed before the year 2085 arrives.)

The spec also says: "After 2084, yy must be 'yy' and only item 4 of this paragraph defines the date."

And this actually made my day 😄!

**Comment by cbrnr** [comment]
Another comment: we don't actually require all dependencies to be on conda-forge, so installing via pip would have also worked. But I guess it doesn't hurt to have it available over there as well.

**Comment by cbrnr** [comment]
There's a problem: edfio requires Python >= 3.9, but MNE currently still supports Python 3.8.

**Comment by hoechenberger** [comment]
Mind you our installers are often lagging behind by 2 release or so, because of the complicated mix of dependencies. Otherwise I'm all for cutting off things that are holding us back.

**Comment by hofaflo** [comment]
Is there already a timeline for dropping Python 3.8 support? If not – or if the plan is not to do that anytime soon – it should be relatively simple to support 3.8 in edfio. Let me know if that would be your preferred solution!


A few more things are fixed now:
- The dictionary mapping integers indicating sex in mne to letters used in EDF uses the mapping given [here](https://mne.tools/stable/generated/mne.Info.html#mne.Info)
- Empty subfields in patient/recording identification are substituted with `"X"` to enusre compatibility with EDF+
- Tolerances for comparing signal values and sample times to their values before exporting have been decreased to make passing the respective assertions non-trivial.
- Slicing signals to their original length has been removed where no padding occurs in the first place.

**Comment by cbrnr** [comment]
I'd prefer that we drop support for Python 3.8. This should be no problem, and we don't have to go full in and also drop 3.9. @drammock @larsoner OK?

**Comment by larsoner** [comment]
We try to follow the scientific Python community standards for this and I think that means it's okay to drop 3.8 according to a quick look at https://numpy.org/neps/nep-0029-deprecation_policy.html#support-table

Let's push 1.6 out the door Monday or Tuesday (it's overdue) then drop 3.8 it while bumping dependency versions to 2-years-old. In the meantime you can ignore the "old" 3.8 CI

**Comment by cbrnr** [comment]
> There's a problem: edfio requires Python >= 3.9, but MNE currently still supports Python 3.8.
> 
> This should be relatively easy to fix though by dropping support for Python 3.8. Actually, since we're already adopting SPECS, [SPEC 0](https://scientific-python.org/specs/spec-0000/) says we should drop 3.8 *and* 3.9 (and support only 3.10, 3.11, and 3.12).

SPEC0 is even more aggressive, but I'm fine with dropping only 3.8 after the 1.6 release.

File: mne/export/_edf.py
Comment (hofaflo) [reply mne/export/_edf.py:27]: Yes, that would be consistent with the EDF specs and maximize the resolution. I wasn't sure if this is an intended deviation from the specs as it allows to trivially set physical min and max so that `0` stays `0` after export and re-import :thinking:

File: mne/export/_edf.py
Comment (hofaflo) [reply mne/export/_edf.py]: Right, I was using them during development and missed to take that one out, thanks!

File: mne/export/_edf.py
Comment (cbrnr) [reply mne/export/_edf.py:27]: I don't know if this was intentional, but if you think that it is a good idea to keep the symmetry, losing one value of resolution is not a big deal.

**Comment by cbrnr** [comment]
One more thing before I'm happy: exporting stimulus channels are not supported, right? For example, [this test](https://github.com/mne-tools/mne-python/blob/main/mne/export/tests/test_export.py#L126) involves a `Raw` object with a stim channel, but it doesn't get exported. If this is intentional (I guess it is, but maybe we can take another look in another PR), I think we should at least emit a warning.

File: mne/export/_edf.py
Comment (hofaflo) [reply mne/export/_edf.py:27]: Yes I would keep this as-is, to avoid breaking workflows relying on that behavior.

**Comment by hofaflo** [comment]
Good idea! Added a warning and removed the stim channel from all but one test to avoid checking for that warning in unrelated tests.

**Comment by cbrnr** [comment]
Great! Can you maybe remove [this comment](https://github.com/hofaflo/mne-python/blob/pr/hofaflo/12218/mne/export/tests/test_export.py#L181) (`# check channel types except for 'bio', which loses its type`), which is incorrect (`bio` stays `bio`)?

This test is a bit convoluted, feel free to clean up a bit if it makes sense. For example, there's actually a third export (the very last line), which I'm not sure what it actually tests (just a smoke test, but for what). And the test is called `test_double_export_edf()`, which is not true because of the third export...

**Comment by Teuniz** [comment]
> One more thing before I'm happy: exporting stimulus channels are not supported, right? For example, [this test](https://github.com/mne-tools/mne-python/blob/main/mne/export/tests/test_export.py#L126) involves a `Raw` object with a stim channel, but it doesn't get exported. If this is intentional (I guess it is, but maybe we can take another look in another PR), I think we should at least emit a warning.

Because I'm curious, what's different about a stim channel that it cannot be saved in EDF+?

**Comment by cbrnr** [comment]
> Because I'm curious, what's different about a stim channel that it cannot be saved in EDF+?

Good question, there's no reason why this cannot be done. In fact, this would be a perfectly valid and useful thing to do, but for some reason MNE-Python has excluded stim channels. We did not want to mess with that in this PR, but we'll tackle it in another PR. I guess the main difference between "normal" (physiological) and stim channels is that the latter contain exact (integer) values, but this can be dealt with by setting the physical and digital minima/maxima appropriately.

**Comment by hofaflo** [comment]
Deconvoluted the tests and rebased onto main, ready for re-review :)

**Review by cbrnr — APPROVED** [review]
Thanks a lot @hofaflo, the tests are now so much more readable! @larsoner please feel free to merge if you're happy.

File: mne/export/_edf.py
Comment (hoechenberger) [inline mne/export/_edf.py:79]: Are we okay with using the walrus operator?

**Comment by hoechenberger** [comment]
I think we'll need a changelog entry :)

File: mne/export/_edf.py
Comment (cbrnr) [reply mne/export/_edf.py:79]: Yes!

Comment (cbrnr) [reply mne/export/_edf.py:79]: But regarding this condition, would it maybe be better to make the comparison more explicit @hofaflo?

Comment (hofaflo) [reply mne/export/_edf.py:79]: I don' really have a preference here, so I'd like to let you decide (also for the [second walrus](https://github.com/hofaflo/mne-python/blob/edfio/mne/export/_edf.py#L199))!

Comment (hoechenberger) [reply mne/export/_edf.py:79]: > But regarding this condition, would it maybe be better to make the comparison more explicit @hofaflo?

Yes that's what I meant, I don't think it helps readability in this specific case. But I'll let you decide

Comment (cbrnr) [reply mne/export/_edf.py:79]: I'd keep it, but I'd make the comparison more explicit by adding `> 0`.

Comment (cbrnr) [reply mne/export/_edf.py:79]: Same for the second one, what exactly are we comparing here? `if meas_date := raw.info["meas_date"] is not None:` maybe?

Comment (hofaflo) [reply mne/export/_edf.py:79]: > the [second walrus](https://github.com/hofaflo/mne-python/blob/edfio/mne/export/_edf.py#L199)

and leave this as-is?

Comment (hoechenberger) [reply mne/export/_edf.py:79]: +1 for @cbrnr's suggestions

**Comment by hofaflo** [comment]
Done!

**Comment by cbrnr** [comment]
Perfect, thanks!

File: mne/export/tests/test_export.py
Comment (larsoner) [inline mne/export/tests/test_export.py:147]: A simpler pattern is just to do:
```
edfio_mark = pytest.mark.skipif(...)
```
then each function can just be decorated with
```
@edfio_mark
def test...
```

File: doc/changes/devel.rst
Comment (larsoner) [inline doc/changes/devel.rst:26]: Should mention the switch from `EDFlib-Python` to `edfio` explicitly because people might need to `pip install` stuff to adjust

**Review by larsoner — COMMENTED** [review]
Just minor stuff from me, happy to defer to @cbrnr's EDF expertise so if he's happy I think we can merge 👍 

You'll want to merge in `main` shortly, though, because @cbrnr dropped 3.8 support in https://github.com/mne-tools/mne-python/pull/12229 which is about to land, and I had to adjust the merge requirements to adapt (and merging `main` into your branch or rebasing will get you the correct CI names/runs to fulfill merge requirements)

**Comment by cbrnr** [comment]
I agree with your comments, so let's wait until they are addressed.

BTW, will this land in 1.6.1 (backported), or do we wait until 1.7? I'd prefer backporting.

**Comment by hofaflo** [comment]
Done, thanks for the suggestions @larsoner!

Backporting to 1.6.1 would require support for Python3.8 in edfio, right?

**Comment by cbrnr** [comment]
> Backporting to 1.6.1 would require support for Python3.8 in edfio, right?

True, I forgot. So I guess this will have to wait for v1.7 then (about 3 months).

**Comment by hoechenberger** [comment]
We should have na "Dependencies" section in the changelog to mention changes of dependencies. @larsoner WDYT?

**Comment by cbrnr** [comment]
Thanks @hofaflo!

**Pull request merged**
